Add Sentry Support

jkmassel opened the pull request:
Adds support for using Sentry via Tracks.

Most of the logic is over in https://github.com/Automattic/Automattic-Tracks-iOS/pull/103. Once that's merged, I'll update this PR with a real version number, instead of a `pod` branch.

**To Test**

1. First off, we'll need to make a couple of small changes to enable testing this:
- Comment out the `BuildConfiguration.is(.debug)` line in [SPPrivacyViewController.swift](https://github.com/Automattic/simplenote-ios/pull/304/files#diff-0893f721a13d096d145c51f82fe2c2b7R19). We want the test crash code to be available no matter what for our testing.
- Change the app's build configuration to Release (edit Scheme > Run > Build Configuration => Release)
- Run `bundle exec fastlane run configure_apply` in the root of the project in order to download the proper configuration.
2. Run the app on-device.
3. Disconnect the device from the host machine, and force-quit the app.
3. Relaunch the app and Log in, ensuring that your user is set up to send crash reports.
4. Navigate to Settings > Privacy Settings > Send a Test Crash (the app will crash)
4. Re-run the app, and let it idle for a moment (this'll ensure the crash report goes to Sentry)
5. Note your shiny new crash event in Sentry – you can see it under "Events" if you filter by `simplenote-ios`.

jleandroperez commented on Podfile:
```diff
@@ -24,12 +24,9 @@ abstract_target 'Automattic' do
-		#
-		pod 'Automattic-Tracks-iOS', '0.3.4'
+		pod 'Automattic-Tracks-iOS', :git => 'https://github.com/Automattic/Automattic-Tracks-iOS.git', :branch => 'add/support-for-tracking-crashes'
 		pod 'Simperium', '0.8.19'
```
This caught my eye (although it wasn't modified in your PR). Latest Simperium release was 0.8.21, just by chance... did you base this branch from another (really old) branch?

bummytime replied:
@jleandroperez nice catch — this branch was cut before `0.8.21` was merged into `develop`.

jkmassel replied:
Haha we actually just merged in `0.8.21` last week: https://github.com/Automattic/simplenote-ios/commit/662769b49c2ba065f0fe0e4c72a238e4da932e5c.

This branch has been rebased on `develop`, and should be good to go.

One more testing step: please be sure to run `bundle exec fastlane configure_apply` before trying to run this, or it'll fail to find the Sentry DSN and things won't work right :)

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+        CrashLogging.start(withDataProvider: dataProvider)
+    }
+
+    @objc static var userHasOptedOut: Bool {
```
Ah, yep. Was used for development debugging.

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+        var emailAddress: String?
+        var didOptOut: Bool = true
+
+        static var empty = User(emailAddress: nil, didOptOut: true)
```
Yep, that's a nice cleanup

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+        }
+    }
+
+    fileprivate static var cachedUser: User? {
```
Ah, yes – I always forget you can do that. Good suggestion :)

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+import AutomatticTracks
+
+/// This exists to bridge CrashLogging with Objective-C. Once the App Delegate is moved over to Swift,
+/// this shim can be deleted.
```
Yeah, my thinking was you could move the caching settings into an extension without too much fuss since they're `static`. I thought about doing it and literally just having the shim bridge Obj-C to Swift, but that seemed like a lot of boilerplate. 

I'll update the comment to better describe exactly what should happen to that code.

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+
+ SNCrashLoggingDataProvider will use this cache in `currentUser` to fetch data on the user.
+ */
+struct CrashLoggingCache {
```
Ah yep, good call. No reason to pollute the project namespace!

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,125 @@
+    }
+}
+
+class SNCrashLoggingDataProvider: CrashLoggingDataProvider {
```
I made the `simperium` property private, and added another inline note for a future developer – when the shim goes, that class will become the main API exposed to the rest of the project.

jkmassel commented:
Thanks @shiki – addressed all of your suggestions in a3da78495ec532146bcfa1ca47323378c3145e11.

>I initially thought that we can make CrashLoggingCache part of SNCrashLoggingDataProvider but that seems to produce a bloated class.

I'm glad I'm not the only one who struggled with that! 😜 

> If only we could add another Simperium.delegate then we could hide the caching inside CrashLoggingShim.

I know I saw a `MultiDelegate` implementation somewhere recently 🤔 😆

shiki commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,123 @@
+import Foundation
+import AutomatticTracks
+
+/// This exists to bridge CrashLogging with Objective-C. Once the App Delegate is moved over to Swift,
```
👌

shiki commented on Simplenote/CrashLogging.swift:
```diff
@@ -110,6 +107,7 @@ struct CrashLoggingCache {
         }
     }
 
+    private(set)
```
You can remove the `fileprivate` too. 😉 

```swift 
    private(set) static var cachedUser: User? {
```

jkmassel replied:
Sure can! 1eb9d81d7162245a7bbcad587bd57a7609c97f3b

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,123 @@
+            return cachedUser?.emailAddress
+        }
+        set {
+            var updatedUser = cachedUser ?? User.empty
```
I like that from a simplification perspective! Implemented in 8cafa87e38ec8ed8dcd6187eaac96c8ea1d2eee9.

jkmassel commented on Simplenote/CrashLogging.swift:
```diff
@@ -0,0 +1,123 @@
+    private(set)
+    fileprivate static var cachedUser: User? {
+        get {
+            guard let data = UserDefaults.standard.object(forKey: key) as? Data else { return nil }
```
We must respect the standards! fb3f3080fe928707b44ad4fe401c66ee564d9dea

jkmassel commented on Simplenote/SPAppDelegate.m:
```diff
@@ -644,6 +630,11 @@ - (void)simperiumDidLogin:(Simperium *)simperium
+    // Now that the user info is present, cache it for use by the crash logging system.
+    // See the docs there for details on why this is necessary.
+    [CrashLogging cacheUser:simperium.user];
+    [CrashLogging cacheOptOutSetting: !simperium.preferencesObject.analytics_enabled.boolValue];
```
👌💯 

84fbdacb50e9175a92111acca0a114301eae9593
